client/my-sites/post-type-list/index.jsx

```diff
@@ -11,52 +18,171 @@ import QueryPosts from 'components/data/query-posts';
+		requestingLastPage: PropTypes.bool
+	};
+
+	constructor() {
+		super( ...arguments );
+
+		this.renderPostRow = this.renderPostRow.bind( this );
+		this.setRequestedPages = this.setRequestedPages.bind( this );
+		this.setVirtualScrollRef = this.setVirtualScrollRef.bind( this );
+		this.state = { requestedPages: [] };
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( ! isEqual( this.props.query, nextProps.query ) ) {
+			this.setState( { requestedPages: [] } );
+		}
+	}
+
+	componentDidUpdate( prevProps ) {
+		// If, after requesting, it turns out that there's only a single post,
+		// VirtualScroll may not update because the row count doesn't change
+		// (1 placeholder, 1 post result). In these cases, force an update.
+		if ( this.props.posts && 1 === this.props.posts.length && this.virtualScroll &&
+				! this.props.requestingFirstPage && prevProps.requestingFirstPage ) {
+			this.virtualScroll.forceUpdate();
+		}
+	}
+
+	setVirtualScrollRef( ref ) {
+		this.virtualScroll = ref;
+	}
+
+	getPageForIndex( index ) {
+		const { query, lastPage } = this.props;
+		const perPage = query.number || DEFAULT_POSTS_PER_PAGE;
+		const page = Math.ceil( index / perPage );
+
+		return Math.max( Math.min( page, lastPage || Infinity ), 1 );
+	}
+
+	setRequestedPages( { startIndex, stopIndex } ) {
+		if ( ! this.props.query ) {
+			return;
+		}
+
+		const { requestedPages } = this.state;
+		const pagesToRequest = difference( range(
+			this.getPageForIndex( startIndex - LOAD_OFFSET ),
+			this.getPageForIndex( stopIndex + LOAD_OFFSET ) + 1
+		), requestedPages );
+
+		if ( ! pagesToRequest.length ) {
+			return;
+		}
+
+		this.setState( {
+			requestedPages: requestedPages.concat( pagesToRequest )
+		} );
+	}
+
+	isLastPage() {
+		const { lastPage, requestingLastPage } = this.props;
+		const { requestedPages } = this.state;
+
+		return includes( requestedPages, lastPage ) && ! requestingLastPage;
+	}
+
+	getRowCount() {
+		let count = 1;
+
+		if ( this.props.posts ) {
+			count += this.props.posts.length;
+		}
+
+		if ( this.isLastPage() ) {
+			count--;
+		}
+
+		return count;
+	}
+
+	renderPostRow( { index } ) {
+		const { requestingFirstPage, posts } = this.props;
+		if ( ! posts || ( requestingFirstPage && 0 === index ) ||
+				( ! requestingFirstPage && index >= posts.length ) ) {
+			return <PostTypeListPostPlaceholder key="placeholder" />;
+		}
+
+		const { global_ID: globalId } = posts[ requestingFirstPage ? index - 1 : index ];
+		return <PostTypeListPost key={ globalId } globalId={ globalId } />;
+	}
+
+	render() {
+		const { query, siteId, posts } = this.props;
+		const isEmpty = query && posts && ! posts.length && this.isLastPage();
+
+		return (
+			<div className="post-type-list">
+				{ query && this.state.requestedPages.map( ( page ) => (
+					<QueryPosts
+						key={ `query-${ page }` }
+						siteId={ siteId }
+						query={ { ...query, page } } />
 				) ) }
-			</ul>
-		</div>
-	);
+				{ isEmpty && (
```

Comment:
While working on the `TermSelector` I discovered the `noRowsRenderer` option for `VirtualScroll`.  Not sure if you want to leverage that here to show `PostTypeListEmptyContent` and/or loading state.